x/sigs/handler_test.go

```diff
@@ -0,0 +1,173 @@
+				key2.Condition(), // Main signer.
+				key1.Condition(),
+			},
+			Msg: BumpSequenceMsg{Increment: 2},
+			WantSequences: []*UserData{
+				{Pubkey: key1, Sequence: 1},
+				{Pubkey: key2, Sequence: 10},
+			},
+		},
+		"transaction with a missing signature is rejected": {
+			Msg:          BumpSequenceMsg{Increment: 1},
+			Signers:      nil,
+			WantCheckErr: errors.ErrUnauthorized,
+		},
+		"message with a zero sequence increment is invalid": {
+			InitData: []*UserData{
+				{Pubkey: key1, Sequence: 1},
+			},
+			Msg:          BumpSequenceMsg{Increment: 0},
+			WantCheckErr: errors.ErrInvalidMsg,
+		},
+		"user that we increment the sequence of must exist": {
+			InitData: []*UserData{
+				{Pubkey: key2, Sequence: 4},
+			},
+			Signers:      []weave.Condition{key1.Condition()},
+			Msg:          BumpSequenceMsg{Increment: 421},
+			WantCheckErr: errors.ErrNotFound,
+		},
+		"sequence increment value must not be greater than 1000": {
+			InitData: []*UserData{
+				{Pubkey: key1, Sequence: 4},
+			},
+			Signers:      []weave.Condition{key1.Condition()},
+			Msg:          BumpSequenceMsg{Increment: 1001},
+			WantCheckErr: errors.ErrInvalidMsg,
+		},
+		"sequence increment value can be 1000": {
+			InitData: []*UserData{
+				{Pubkey: key1, Sequence: 4},
+			},
+			Signers: []weave.Condition{key1.Condition()},
+			Msg:     BumpSequenceMsg{Increment: 1000},
+			WantSequences: []*UserData{
+				{Pubkey: key1, Sequence: 1003},
+			},
+		},
+		"successfull sequence increment before counter overflow": {
+			InitData: []*UserData{
+				{Pubkey: key1, Sequence: math.MaxInt64 - 20},
+			},
+			Signers: []weave.Condition{key1.Condition()},
+			Msg:     BumpSequenceMsg{Increment: 20},
+			WantSequences: []*UserData{
+				{Pubkey: key1, Sequence: math.MaxInt64 - 1},
+			},
+		},
+		"sequence increment value overflow": {
+			InitData: []*UserData{
+				{Pubkey: key1, Sequence: math.MaxInt64 - 20},
+			},
+			Signers:      []weave.Condition{key1.Condition()},
+			Msg:          BumpSequenceMsg{Increment: 21},
+			WantCheckErr: errors.ErrOverflow,
+		},
+	}
+
+	for testName, tc := range cases {
+		t.Run(testName, func(t *testing.T) {
+			bucket := NewBucket()
+			db := store.MemStore()
+
+			for i, data := range tc.InitData {
+				obj := orm.NewSimpleObj(data.Pubkey.Address(), data)
+				if err := bucket.Save(db, obj); err != nil {
+					t.Fatalf("cannot save %d user: %s", i, err)
+				}
+			}
+
+			auth := &weavetest.CtxAuth{Key: "auth"}
+			handler := bumpSequenceHandler{
+				b:    bucket,
+				auth: auth,
+			}
+			ctx := context.Background()
+			ctx = auth.SetConditions(ctx, tc.Signers...)
+			tx := weavetest.Tx{Msg: &tc.Msg}
+
+			cache := db.CacheWrap()
+			if _, err := handler.Check(ctx, cache, &tx); !tc.WantCheckErr.Is(err) {
+				t.Fatalf("unexpected check error: %+v", err)
+			}
+			cache.Discard()
+
+			if tc.WantCheckErr != nil {
+				// If we expect an error than it makes no sense to continue the flow.
+				return
```

Comment:
pp: you are using this pattern in your tests where you don't execute delivery when check has already passed. It would be more efficient when you fall through and recheck the conditions on deliver as well. The motivation for this is that the TX order in a block can create a different state than was on check in the past. You don't have confidence  **from the tests** that the validation code is executed on deliver as well.